### ISSUE TITLE
Migrations and 7.0.0 don't play nicely

### ISSUE THREAD

**Issue by tmtrademarked** [realm/realm-java#6867]
#### Goal
Upgrade from 6.1.0 to 7.0.0 - without a crash. I am seeing a crash after running migrations in 7.0.0, and it is maddening to chase down - the same migrations and schema work 100% OK in 6.1.0, but explode in 7.0.0 with a JNI exception.

#### Actual Results

I have an instrumentation test which we use to verify migrations from old schemas into our new schema. When I updated to realm-java 7.0.0 from 6.1.0, my test started failing:
```
05-19 23:01:01.722 D/RootInstrumentationTest(20518): >>>>>>>>>>>> Starting test: runAllMigrations(com.blueapron.service.cache.RealmGateTest)
05-19 23:01:02.322 W/on.service.tes(20518): Accessing hidden method Ldalvik/system/BaseDexClassLoader;->addDexPath(Ljava/lang/String;Z)V (greylist, reflection, allowed)
05-19 23:01:02.548 W/on.service.tes(20518): Accessing hidden method Lsun/misc/Unsafe;->allocateInstance(Ljava/lang/Class;)Ljava/lang/Object; (greylist, linking, allowed)
05-19 23:01:02.548 W/on.service.tes(20518): Accessing hidden field Lsun/misc/Unsafe;->theUnsafe:Lsun/misc/Unsafe; (greylist, reflection, allowed)
05-19 23:01:03.258 D/AnalyticsReporter(20518): Logging event: Diagnostic - Schema Migration - Started - M. Properties: {"initial_version":"72","api_level":"29","diagnostic_event":"true","output_version":"83"}
05-19 23:01:03.267 D/AnalyticsReporter(20518): Logging event: Diagnostic - Schema Migration - Finished - M. Properties: {"initial_version":"72","api_level":"29","diagnostic_event":"true","output_version":"83"}
05-19 23:01:03.269 E/REALM_JNI(20518): jni: ThrowingException 8, Index out of range, .
05-19 23:01:03.269 E/REALM_JNI(20518): Exception has been thrown: Index out of range
05-19 23:01:03.276 W/System.err(20518): FLAKE_FAILURE: runAllMigrations(com.blueapron.service.cache.RealmGateTest) failed on run 1
05-19 23:01:03.277 W/System.err(20518): java.lang.IllegalStateException: Index out of range
05-19 23:01:03.277 W/System.err(20518): 	at io.realm.internal.OsSharedRealm.nativeGetSharedRealm(Native Method)
05-19 23:01:03.277 W/System.err(20518): 	at io.realm.internal.OsSharedRealm.<init>(OsSharedRealm.java:175)
05-19 23:01:03.277 W/System.err(20518): 	at io.realm.internal.OsSharedRealm.getInstance(OsSharedRealm.java:251)
05-19 23:01:03.277 W/System.err(20518): 	at io.realm.BaseRealm.<init>(BaseRealm.java:137)
05-19 23:01:03.277 W/System.err(20518): 	at io.realm.BaseRealm.<init>(BaseRealm.java:104)
05-19 23:01:03.277 W/System.err(20518): 	at io.realm.Realm.<init>(Realm.java:163)
05-19 23:01:03.277 W/System.err(20518): 	at io.realm.Realm.createInstance(Realm.java:499)
05-19 23:01:03.277 W/System.err(20518): 	at io.realm.RealmCache.createInstance(RealmCache.java:507)
05-19 23:01:03.277 W/System.err(20518): 	at io.realm.RealmCache.doCreateRealmOrGetFromCache(RealmCache.java:473)
05-19 23:01:03.277 W/System.err(20518): 	at io.realm.RealmCache.createRealmOrGetFromCache(RealmCache.java:414)
05-19 23:01:03.277 W/System.err(20518): 	at io.realm.Realm.getInstance(Realm.java:428)
05-19 23:01:03.277 W/System.err(20518): 	at com.blueapron.service.cache.RealmGateTest.runAllMigrations(RealmGateTest.java:157)
05-19 23:01:03.277 W/System.err(20518): 	at java.lang.reflect.Method.invoke(Native Method)
05-19 23:01:03.277 W/System.err(20518): 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
05-19 23:01:03.277 W/System.err(20518): 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
05-19 23:01:03.277 W/System.err(20518): 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
05-19 23:01:03.277 W/System.err(20518): 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
05-19 23:01:03.277 W/System.err(20518): 	at androidx.test.internal.runner.junit4.statement.RunBefores.evaluate(RunBefores.java:80)
05-19 23:01:03.277 W/System.err(20518): 	at androidx.test.internal.runner.junit4.statement.RunAfters.evaluate(RunAfters.java:61)
05-19 23:01:03.277 W/System.err(20518): 	at com.blueapron.testcommon.FlakyTestRule$1.evaluate(FlakyTestRule.java:30)
05-19 23:01:03.277 W/System.err(20518): 	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
05-19 23:01:03.277 W/System.err(20518): 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
05-19 23:01:03.277 W/System.err(20518): 	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
05-19 23:01:03.277 W/System.err(20518): 	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
05-19 23:01:03.277 W/System.err(20518): 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
05-19 23:01:03.277 W/System.err(20518): 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
05-19 23:01:03.277 W/System.err(20518): 	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
05-19 23:01:03.277 W/System.err(20518): 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
05-19 23:01:03.277 W/System.err(20518): 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
05-19 23:01:03.277 W/System.err(20518): 	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
05-19 23:01:03.277 W/System.err(20518): 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
05-19 23:01:03.277 W/System.err(20518): 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
05-19 23:01:03.277 W/System.err(20518): 	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
05-19 23:01:03.277 W/System.err(20518): 	at androidx.test.ext.junit.runners.AndroidJUnit4.run(AndroidJUnit4.java:104)
05-19 23:01:03.277 W/System.err(20518): 	at org.junit.runners.Suite.runChild(Suite.java:128)
05-19 23:01:03.277 W/System.err(20518): 	at org.junit.runners.Suite.runChild(Suite.java:27)
05-19 23:01:03.277 W/System.err(20518): 	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
05-19 23:01:03.277 W/System.err(20518): 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
05-19 23:01:03.277 W/System.err(20518): 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
05-19 23:01:03.277 W/System.err(20518): 	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
05-19 23:01:03.277 W/System.err(20518): 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
05-19 23:01:03.277 W/System.err(20518): 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
05-19 23:01:03.277 W/System.err(20518): 	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
05-19 23:01:03.277 W/System.err(20518): 	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
05-19 23:01:03.277 W/System.err(20518): 	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
05-19 23:01:03.277 W/System.err(20518): 	at androidx.test.internal.runner.TestExecutor.execute(TestExecutor.java:56)
05-19 23:01:03.277 W/System.err(20518): 	at androidx.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:388)
05-19 23:01:03.277 W/System.err(20518): 	at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:2189)
05-19 23:01:03.277 W/System.err(20518): runAllMigrations(com.blueapron.service.cache.RealmGateTest) failed after 1 attempts
05-19 23:01:03.277 W/System.err(20518): HARD_ERROR: Test did not succeed after multiple attempts
05-19 23:01:03.278 D/RootInstrumentationTest(20518): <<<<<<<<<<<< Finishing test: runAllMigrations(com.blueapron.service.cache.RealmGateTest)
```

Similarly, if I install these relevant builds on a device, I get the same crash:
```
05-19 23:45:48.779 W/on.service.tes( 9200): Accessing hidden method Lsun/misc/Unsafe;->allocateInstance(Ljava/lang/Class;)Ljava/lang/Object; (greylist, linking, allowed)
05-19 23:45:48.779 W/on.service.tes( 9200): Accessing hidden field Lsun/misc/Unsafe;->theUnsafe:Lsun/misc/Unsafe; (greylist, reflection, allowed)
05-19 23:45:49.050 D/AnalyticsReporter( 9232): Logging event: Diagnostic - Schema Migration - Started - M. Properties: {"initial_version":"72","api_level":"29","diagnostic_event":"true","output_version":"83"}
05-19 23:45:49.144 D/AnalyticsReporter( 9232): Logging event: Diagnostic - Schema Migration - Finished - M. Properties: {"initial_version":"72","api_level":"29","diagnostic_event":"true","output_version":"83"}
05-19 23:45:49.149 E/REALM_JNI( 9232): jni: ThrowingException 8, Index out of range, .
05-19 23:45:49.149 E/REALM_JNI( 9232): Exception has been thrown: Index out of range
05-19 23:45:49.150 D/AndroidRuntime( 9232): Shutting down VM
05-19 23:45:49.150 E/AndroidRuntime( 9232): FATAL EXCEPTION: main
05-19 23:45:49.150 E/AndroidRuntime( 9232): Process: com.blueapron.blueapron.debug, PID: 9232
05-19 23:45:49.150 E/AndroidRuntime( 9232): java.lang.RuntimeException: Unable to create application com.blueapron.mobile.ui.MobileApplication: java.lang.IllegalStateException: Index out of range
05-19 23:45:49.150 E/AndroidRuntime( 9232): 	at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6465)
05-19 23:45:49.150 E/AndroidRuntime( 9232): 	at android.app.ActivityThread.access$1300(ActivityThread.java:219)
05-19 23:45:49.150 E/AndroidRuntime( 9232): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1859)
05-19 23:45:49.150 E/AndroidRuntime( 9232): 	at android.os.Handler.dispatchMessage(Handler.java:107)
05-19 23:45:49.150 E/AndroidRuntime( 9232): 	at android.os.Looper.loop(Looper.java:214)
05-19 23:45:49.150 E/AndroidRuntime( 9232): 	at android.app.ActivityThread.main(ActivityThread.java:7356)
05-19 23:45:49.150 E/AndroidRuntime( 9232): 	at java.lang.reflect.Method.invoke(Native Method)
05-19 23:45:49.150 E/AndroidRuntime( 9232): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
05-19 23:45:49.150 E/AndroidRuntime( 9232): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
05-19 23:45:49.150 E/AndroidRuntime( 9232): Caused by: java.lang.IllegalStateException: Index out of range
05-19 23:45:49.150 E/AndroidRuntime( 9232): 	at io.realm.internal.OsSharedRealm.nativeGetSharedRealm(Native Method)
05-19 23:45:49.150 E/AndroidRuntime( 9232): 	at io.realm.internal.OsSharedRealm.<init>(OsSharedRealm.java:175)
05-19 23:45:49.150 E/AndroidRuntime( 9232): 	at io.realm.internal.OsSharedRealm.getInstance(OsSharedRealm.java:251)
05-19 23:45:49.150 E/AndroidRuntime( 9232): 	at io.realm.BaseRealm.<init>(BaseRealm.java:137)
05-19 23:45:49.150 E/AndroidRuntime( 9232): 	at io.realm.BaseRealm.<init>(BaseRealm.java:104)
05-19 23:45:49.150 E/AndroidRuntime( 9232): 	at io.realm.Realm.<init>(Realm.java:163)
05-19 23:45:49.150 E/AndroidRuntime( 9232): 	at io.realm.Realm.createInstance(Realm.java:499)
05-19 23:45:49.150 E/AndroidRuntime( 9232): 	at io.realm.RealmCache.createInstance(RealmCache.java:507)
05-19 23:45:49.150 E/AndroidRuntime( 9232): 	at io.realm.RealmCache.doCreateRealmOrGetFromCache(RealmCache.java:473)
05-19 23:45:49.150 E/AndroidRuntime( 9232): 	at io.realm.RealmCache.createRealmOrGetFromCache(RealmCache.java:414)
05-19 23:45:49.150 E/AndroidRuntime( 9232): 	at io.realm.Realm.getDefaultInstance(Realm.java:407)
05-19 23:45:49.150 E/AndroidRuntime( 9232): 	at com.blueapron.service.cache.RealmGate.open(RealmGate.kt:58)
05-19 23:45:49.150 E/AndroidRuntime( 9232): 	at com.blueapron.service.analytics.AnalyticsReporter.initialize(AnalyticsReporter.java:156)
05-19 23:45:49.150 E/AndroidRuntime( 9232): 	at com.blueapron.service.ui.BlueApronApplication.onCreate(BlueApronApplication.java:95)
05-19 23:45:49.150 E/AndroidRuntime( 9232): 	at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1182)
05-19 23:45:49.150 E/AndroidRuntime( 9232): 	at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6460)
05-19 23:45:49.150 E/AndroidRuntime( 9232): 	... 8 more
05-19 23:45:49.303 D/FA      ( 9232): Event not sent since app measurement is disabled
```

If, instead, I keep Realm at 6.1.0, the same test passes flawlessly, and the app opens just fine. The only difference between builds that have the error and builds that do not is the version of Realm we're using.

#### Steps & Code to Reproduce

I don't quite know how to provide a sample or a cleaner test case here. It *seems* to only be happening for me when upgrading from certain versions of our past schema, but it's not 100% obvious why or which migrations produce problems - or even if it *is* migration related. It *seems* like it's a migration issue, because the crash seems to happen between when migrations complete and when the Realm itself opens - but all indications are that our migration code is completing and handling back control to the Realm native code cleanly.

From several exhausting rounds of debugging, I am pretty confident that *somehow* this is related to a migration - and that it depends on the exact content of the Realm file, since depending on what state is in the Realm when I upgrade, I get different behavior during the upgrade. But at the very least, there's *something* behaving differently in the bowels of Realm's internals that is not getting reported well somehow. :-/

It's possible this is just another manifestation of the bug I reported earlier today (https://github.com/realm/realm-java/issues/6866) that just happens to show up differently based on *exactly* what migration is run with *exactly* what data.

#### Version of Realm and tooling

Realm version(s): 6.1.0 -> 7.0.0

Realm Sync feature enabled: No

Android Studio version: 3.6.3

Gradle version: 6.3

Which Android version and device(s): Emulator Pixel 3, SDK 29

**Comment by cmelchior** [comment]
Hmm, Index out of range seems like a really weird error message when opening a Realm. Can you share the integration test and the Realm file? If you want to keep it private you can send it to christian.melchior@mongodb.com

**Comment by tmtrademarked** [comment]
Replied directly - it's a little tricky to send a fully executable test, but I sent over the relevant source code. If that doesn't provide enough information, I can probably figure out how to package up a fully runnable example to debug further!

**Comment by cmelchior** [comment]
I tried to reproduce the error with the information you provided, but so far no luck :( 
Judging from the error you are getting I suspect it might be related to how we send Realm version data to the native layer, but I'm a bit puzzled exactly what is going wrong.

It sounds like you can reproduce this on your end? Would it be possible for you to attach a debugger and see the values in `OsSharedRealm` here https://github.com/realm/realm-java/blob/master/realm/realm-library/src/main/java/io/realm/internal/OsSharedRealm.java#L175 . I'm particularly interested in values for `version.version` and `version.index`

**Comment by tmtrademarked** [comment]
Here's a screenshot of the values. The `version.version` and `version.index` values are both `-1`, which from reading the code seems to indicate that we requested the live version instead of a frozen one, correct?

<img width="581" alt="Screen Shot 2020-05-26 at 2 17 41 PM" src="https://user-images.githubusercontent.com/16063079/82936184-339dab00-9f5c-11ea-9389-1b8b216ab312.png">

Let me know if there's additional data that would be useful here - and thanks for taking a look!

**Comment by tmtrademarked** [comment]
Hey, folks - just wanted to see if there was anything else that I can provide to assist debugging here. I'm able to reproduce this locally, but thus far haven't been able to create a more minimal sample app that also reproduces.

**Comment by cmelchior** [comment]
We believe this should be fixed in 7.0.1 which was just released. Please reopen if that is not the case.

**Comment by tmtrademarked** [comment]
Great - thanks, @cmelchior ! Our testing so far seems to indicate this release did in fact fix our issues. I'll keep an eye on things over the next week and make sure that stays true.
